README.md
Moo
===
Moo is a minimalist Object Orientation system for Perl 5 that aims to be fully
interoperable with Moose.

  * [Moo on CPAN](https://metacpan.org/pod/Moo) contains the main documentation
  * [Role::Tiny on CPAN](https://metacpan.org/pod/Role::Tiny)
  * [Moose on CPAN](https://metacpan.org/pod/Moose)

--------
[![Travis Build Status](https://api.travis-ci.com/moose/Moo.png?branch=master)](https://travis-ci.com/github/moose/Moo)
[![AppVeyor Build status](https://ci.appveyor.com/api/projects/status/github/moose/Moo?branch=master&svg=true)](https://ci.appveyor.com/project/moose/moo/branch/master)
[![Coverage Status](https://codecov.io/github/moose/Moo/coverage.svg?branch=master)](https://codecov.io/github/moose/Moo?branch=master)
[![CPAN version](https://badge.fury.io/pl/Moo.svg)](https://metacpan.org/pod/Moo)


Changes
Revision history for Moo

2.005005 - 2023-01-05
  - drop dependency on Test::Fatal

2.005004 - 2021-03-29
  - fix file set in %INC in create_class_with_roles (RT#134845)

2.005003 - 2021-03-21
  - silence 'once' warnings from Moo::_Utils

2.005_002 - 2021-01-22
  - fix C3 test skipping properly when MRO::Compat not available on perl < 5.10

2.005_001 - 2021-01-20
  - fix perl version check in global destruction code

2.005_000 - 2021-01-20
  - remove MooseX::Types from developer prereqs
  - recommend Sub::Util rather than Sub::Name, since Sub::Util is in core
  - fix line numbers when using oo module (perl -Moo)
  - adjust some author tests to rely less on external modules
  - lower Exporter prereq to any version
  - bump Role::Tiny prereq to 2.003004
  - refactor and simplify role application code, as allowed by new Role::Tiny
    version
  - switch to using normal strict+warnings rather than strictures for authors
  - remove strictures from recommends, as Moo does not use it anywhere
  - remove Task::Weaken prereq, as it served no purpose
  - remove MRO::Compat prereq, using it only when user code does
  - remove use of Devel::GlobalDestruction in code, only using
    Devel::GlobalDestruction::XS when it is available. Devel::GlobalDestruction
    is still a dependency on perl < 5.14 to facilitate the installation of the
    ::XS module.
  - Moo now has no mandatory perl version specific dependencies. The version
    specific modules are now optional or not used. This means code using Moo can
    be fatpacked on a new perl version without special cases, and it will work
    on older perl versions.

2.004004 - 2020-11-25
  - fix error location test when using core Carp on perl 5.8.9 - 5.10.1

2.004003 - 2020-11-21
  - fix extraneous MANIFEST entry

2.004002 - 2020-11-20
  - no changes
  - releasing as stable

2.004_001 - 2020-07-30
  - remove Module::Runtime prerequisite
  - internal code cleanups
  - added test for conflicts in extensions wrapping 'has' and similar subs

2.004000 - 2020-04-09
  - minor documentation tweaks
  - minor test tweaks
  - refactor creation and installation of helper subs allowing extensions to
    customize the behavior easier.
  - added is_class and make_class methods to Moo
  - added make_role method to Moo::Role
  - lower Scalar::Util dependency to 1.00
  - fix unlikely failure to properly inflate Type::Tiny types to Moose under
    threads on perl 5.8

2.003006 - 2019-10-25
  - update Role::Tiny prerequisite to 2.001004 to fix issues with re-importing
    Moo::Role

2.003_005 - 2019-10-18
  - always exclude helper subs (has, with, etc) from the list of methods, even
    if they are installed later or wrapped with a modifier
  - drop Devel::GlobalDestruction prerequisite unless using perl < 5.14
  - fix preserving full accuracy of numbers in non-ref defaults
  - fix tracking of stubs and constants as methods to be consistent and work
    properly with unusual ways of storing subs (RT#130050)
  - fix test for modules broken by newer Moo versions when new enough
    CPAN::Meta version is not available
  - fix undeferring subs before wrapping with a method modifier when subs to
    wrap are given as a list or array ref
  - fix error reporting locations from Carp calls in DEMOLISH (RT#124845)
  - fix extending attributes (has +) to allow overriding a default with a
    builder (RT#130361)
  - fix re-throwing Moose inflation exceptions when examining exception
    objects on older Moose or Devel::StackTrace versions
  - reorganized documentation related to Moose to improve clarity
  - improved documentation of usage with namespace::clean
  - various documentation tweaks

2.003004 - 2017-12-01
  - re-allow stubs for attribute parameters like isa or coerce (RT#123753)
  - fix accidentally removed space in coderef error message (GH#33)
  - fix test errors with old Carp versions

2.003003 - 2017-11-16
  - test tweaks
  - fix handling of code refs stored directly in the stash (for perl 5.28)
  - consider inline packages with constants in them as being loaded
  - stubs will be treated as methods that exist when inflating to Moose
  - avoid loading overload.pm unless required

2.003002 - 2017-03-28
  - ensure tarball does not contain SCHILY headers

2.003001 - 2017-03-06
  - fix +attributes replacing builder subs if parent attribute was defined with
    builder => $subref
  - fix trigger with a default value and init_arg of undef

2.003000 - 2016-12-09
  - fix create_class_with_roles being used multiple times with the same packages
  - fix edge case with @ISA assignment on perl 5.10.0
  - minor test adjustments
  - fix handles on oddly named attributes
  - make has options linkable in documentation
  - Sub::Quote and Sub::Defer have been split into a separate dist

2.002005 - 2016-10-31
  - fix accessor extensions that need captured variables for clearers and
    predicates. (RT#118453)
  - avoid relying on '.' being in @INC in tests
  - fix Sub::Quote test when run with perl -C or PERL_UNICODE on perl 5.10
    (RT#117844)
  - improved error messages for invalid sub names in Sub::Quote (RT#116416,
    RT#117711)
  - clarify meta method documentation
  - bump Role::Tiny prereq version to get stub in role fix (RT#116674)

2.002004 - 2016-06-28
  - fixed another case of local functions interfering with generated code.
    (RT#115655)
  - prevent infinite recursion on some Moose metaclass inflation errors.

2.002003 - 2016-06-23
  - prevent local functions with same names as core functions from interfering
    with generated code (RT#115529)
  - Work around nmake bug that corrupts commands that include slashes
    (RT#115518)
  - Fix tests to work when lexical features are enabled outside of our control
    (such as with cperl)
  - Fix tests on perl 5.6

2.002002 - 2016-06-21
  - fix handling of Carp < 1.12

2.002_001 - 2016-06-17
  - added Sub::Quote::sanitize_identifier to generate an identifier from an
    arbitrary string.
  - Sub::Defer::defer_info is now exportable.

Makefile.PL
use strict;
use warnings;
use 5.006;

my %META = (
  name => 'Moo',
  license => 'perl_5',
  prereqs => {
    configure => { requires => {
      'ExtUtils::MakeMaker'   => 0,
    } },
    build => { requires => {
    } },
    test => {
      requires => {
        'Test::More'  => '0.94',
      },
      recommends => {
        'Parse::CPAN::Meta'         => '1.4200',
        'CPAN::Meta::Requirements'  => 0,
      },
    },
    runtime => {
      requires => {
        'perl'                      => '5.006',
        'Scalar::Util'              => '1.00',
        'Exporter'                  => 0,
        'Carp'                      => 0,
        'Class::Method::Modifiers'  => '1.10',  # for RT#80194
        'Role::Tiny'                => '2.002003',
        'Sub::Quote'                => '2.006006',
        'Sub::Defer'                => '2.006006',
      },
      recommends => {
        'Class::XSAccessor'         => '1.18',
        'Sub::Util'                 => '0',
      },
    },
    develop => {
      requires => {
        'Class::Tiny'                     => '1.001',
        'Moose'                           => '1.15',
        'Mouse'                           => 0,
        'namespace::autoclean'            => 0,
        'namespace::clean'                => 0,
        'Type::Tiny'                      => '0.004',
      },
    },
  },
  resources => {
    repository => {
      url => 'https://github.com/moose/Moo.git',
      web => 'https://github.com/moose/Moo',
      type => 'git',
    },
    x_IRC => 'irc://irc.perl.org/#moose',
    bugtracker => {
      web => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Moo',
      mailto => 'bug-Moo@rt.cpan.org',
    },
    license => [ 'https://dev.perl.org/licenses/' ],
  },
  no_index => {
    directory => [ 't', 'xt' ]
  },
  x_breaks => {
    'HTML::Restrict'            => '== 2.1.5',
    'MySQL::Workbench::Parser'  => '<= 0.05',
    'MooX::Emulate::Class::Accessor::Fast' => '<= 0.02',
    'WebService::Shutterstock'  => '<= 0.006',
    'File::DataClass'           => '<= 0.54.1',
    'App::Commando'             => '<= 0.012',
  },
  x_authority => 'cpan:MSTROUT',
);

my $xt = $ENV{EXTENDED_TESTING};

my %MM_ARGS = (
  PREREQ_PM => {
    ("$]" < 5.014_000 ? ('Devel::GlobalDestruction' => '0.11') : ()),
  },
  ($xt ? (
    TEST_REQUIRES => {
      %{ $META{prereqs}{runtime}{recommends} },
      %{ $META{prereqs}{test}{recommends} },
      %{ $META{prereqs}{develop}{requires} },
    },
    test => { TESTS => 't/*.t xt/*.t' },
  ):()),
);

{
  package MY;

  sub test_via_harness {
    my($self, $perl, $tests) = @_;
    $perl .= ' -I'.$self->catdir('t','lib').' "-MTestEnv=$(MOO_TEST_ENV)"';
    return $self->SUPER::test_via_harness($perl, $tests);
  }

  sub postamble {
    my $MOO_TEST_ENV = (!-f 'META.yml' || $xt) ? "MOO_FATAL_WARNINGS" : '';
    ($xt ? <<"XT" : '')
test :: test_no_xs
XT
    .<<"POSTAMBLE"
MOO_TEST_ENV=$MOO_TEST_ENV
fulltest: test test_no_xs
\t\$(NOECHO) \$(NOOP)
test_no_xs: \$(TEST_TYPE)_no_xs
\t\$(NOECHO) \$(NOOP)
POSTAMBLE
    .join('', map <<"TEST_TYPE", 'dynamic', 'static', '')
test_${_}_no_xs: pure_all
\t\$(NOECHO)\$(MAKE) test_$_ MOO_TEST_ENV="\$(MOO_TEST_ENV),MOO_XS_DISABLE"
TEST_TYPE
  }
}

## BOILERPLATE ###############################################################
require ExtUtils::MakeMaker;
(do './maint/Makefile.PL.include' or die $@) unless -f 'META.yml';

# have to do this since old EUMM dev releases miss the eval $VERSION line
my $eumm_version  = eval $ExtUtils::MakeMaker::VERSION;
my $mymeta        = $eumm_version >= 6.57_02;
my $mymeta_broken = $mymeta && $eumm_version < 6.57_07;

($MM_ARGS{NAME} = $META{name}) =~ s/-/::/g;
($MM_ARGS{VERSION_FROM} = "lib/$MM_ARGS{NAME}.pm") =~ s{::}{/}g;
$META{license} = [ $META{license} ]
  if $META{license} && !ref $META{license};
$MM_ARGS{LICENSE} = $META{license}[0]
  if $META{license} && $eumm_version >= 6.30;
$MM_ARGS{NO_MYMETA} = 1
  if $mymeta_broken;
$MM_ARGS{META_ADD} = { 'meta-spec' => { version => 2 }, %META }
  unless -f 'META.yml';
$MM_ARGS{PL_FILES} ||= {};
$MM_ARGS{NORECURS} = 1
  if not exists $MM_ARGS{NORECURS};

for (qw(configure build test runtime)) {
  my $key = $_ eq 'runtime' ? 'PREREQ_PM' : uc $_.'_REQUIRES';
  my $r = $MM_ARGS{$key} = {
    %{$META{prereqs}{$_}{requires} || {}},
    %{delete $MM_ARGS{$key} || {}},
  };
  defined $r->{$_} or delete $r->{$_} for keys %$r;
}

$MM_ARGS{MIN_PERL_VERSION} = delete $MM_ARGS{PREREQ_PM}{perl} || 0;

delete $MM_ARGS{MIN_PERL_VERSION}
  if $eumm_version < 6.47_01;
$MM_ARGS{BUILD_REQUIRES} = {%{$MM_ARGS{BUILD_REQUIRES}}, %{delete $MM_ARGS{TEST_REQUIRES}}}
  if $eumm_version < 6.63_03;
$MM_ARGS{PREREQ_PM} = {%{$MM_ARGS{PREREQ_PM}}, %{delete $MM_ARGS{BUILD_REQUIRES}}}
  if $eumm_version < 6.55_01;
delete $MM_ARGS{CONFIGURE_REQUIRES}
  if $eumm_version < 6.51_03;

ExtUtils::MakeMaker::WriteMakefile(%MM_ARGS);
## END BOILERPLATE ###########################################################
